lib/spack/docs/module_file_support.rst

```diff
@@ -0,0 +1,704 @@
+installation of a package. They are placed in the following directories
+under the Spack root:
+
+  +--------------------------------+------------------------------------+
+  |                                |  **Module files root**             |
+  +================================+====================================+
+  |  **Environment Modules**       |  share/spack/modules               |
+  +--------------------------------+------------------------------------+
+  |  **Dotkit**                    |   share/spack/dotkit               |
+  +--------------------------------+------------------------------------+
+  |  **LMod Hierarchical Modules** |   share/spack/lmod                 |
+  +--------------------------------+------------------------------------+
+
+
+Though Spack ships with sensible defaults for the generation of module files,
+one can customize many aspects of it to accommodate package or site specific needs.
+These customizations are enabled by either:
+
+ 1. overriding certain callback APIs in the Python packages
+ 2. writing specific rules in the ``modules.yaml`` configuration file
+
+The former method fits best cases that are site independent, e.g. injecting variables
+from language interpreters into their extensions. The latter instead permits to
+fine tune the content, naming and creation of module files to meet site specific conventions.
+
+--------------------------
+Overriding ``Package`` API
+--------------------------
+
+There are two methods that can be overridden in any ``package.py`` to affect the
+content of generated module files. The first one is:
+
+.. code-block:: python
+
+   def setup_environment(self, spack_env, run_env):
+       """Set up the compile and runtime environments for a package."""
+       pass
+
+and can alter the content of *the same package where it is overridden*
+by adding actions to ``run_env``. The second method is:
+
+.. code-block:: python
+
+   def setup_dependent_environment(self, spack_env, run_env, dependent_spec):
+       """Set up the environment of packages that depend on this one"""
+       pass
+
+and has similar effects on module file of dependees. Even in this case
+``run_env`` must be filled with the desired list of environment modifications.
+
+.. note::
+ The ``R`` package and callback APIs
+  A typical example in which overriding both methods prove to be useful
+  is given by the ``R`` package. This package installs libraries and headers
+  in non-standard locations and it is possible to prepend the appropriate directory
+  to the corresponding environment variables:
+
+  ================== =================================
+   LIBRARY_PATH       ``self.prefix/rlib/R/lib``
+   LD_LIBRARY_PATH    ``self.prefix/rlib/R/lib``
+   CPATH              ``self.prefix/rlib/R/include``
+  ================== =================================
+
+  with the following snippet:
+
+  .. literalinclude:: ../../../var/spack/repos/builtin/packages/R/package.py
+     :lines: 151-159
```

Comment:
Instead of specifying line numbers, which are subject to change, another way you can get what you want is to specify a function:
```rst
.. literalinclude:: ../../../var/spack/repos/builtin/packages/R/package.py
   :pyobject: R.setup_environment
```